98_Drivers/Timers.md

```diff
@@ -0,0 +1,283 @@
+
+Other than setting the initial count, we also have to set up the timer LVT entry. There's a few fields here, but we're mostly interested in the following:
+
+- Bits 7:0: this is interrupt vector the timer will trigger when it expires. It will only trigger that vector on the core the LAPIC is attached to.
+- Bit 16: Acts as a mask bit, if set the timer won't generate an interrupt when expiring.
+- Bits 18:17: The mode field. Set this to 0b00 for one-shot operation, and 0b01 for periodic.
+
+The intel and AMD manuals contain the full description if you want to explore the other functionality offered.
+
+```c
+void arm_lapic_interrupt_timer(size_t millis, uint8_t vector) {
+    volatile uint32_t* lvt_reg = lapic_regs + 0x320;
+    //note this clears bits 16 (mask) and 18:17 (mode)
+    *lvt_reg = vector;
+
+    uint32_t ticks = lapic_ms_to_ticks(millis);
+    volatile uint32_t* init_reg = lapic_regs + 0x380;
+    *init_reg = ticks;
+}
+```
+
+## Timestamp Counter (TSC)
+
+The TSC is a bit more modern than the LAPIC timer, but still pre-dates most long mode processors, so this is another timer that should always be present. Having said that, it can be checked for using cpuid leaf 1, edx bit 4. 
+
+The TSC is probably the simplest timer we've covered so far: it's simply a 64-bit counter that increments every time the base clock of the processor pulses. To read this counter we can use the `rdtsc` instruction which places the low 33-bits in eax and high 32-bits in edx. Similar to how the MSR instructions work.
+
+There are some issues with this version of the TSC however: modern processors will change their base speed depending on power/performance requirements, which means that the rate the TSC ticks at will change dynamically! This makes it pretty useless as a timer, and a newer version was quickly implemented, called the invariant TSC.
+
+The I-TSC ticks at the base speed the processor is supposed to run at, not what it's actually running at, meaning the tick-rate is constant. Most processors support the I-TSC nowdays, and most emulators also do, even if they dont advertise it through cpuid (qemu has invariant TSC, but doesn't set the bit). To test if the TSC is invariant can be done via cpuid once again: leaf 7, edx bit 8.
+
+How about generating interrupts with the TSC? This is also an option feature (that's almost always supported) called TSC deadline. We can test for it's existence via cpuid leaf 1, ecx, bit 24. To use TSC deadline we write the absolute time (in TSC ticks) of when we want the interrupt to a special MSR, called IA_32_TSC_DEADLINE (MSR 0x6E0).
+
+When the TSC passes the tick value in this MSR, it tells the local APIC, and if TSC deadline mode is selected in the timer LVT an interrupt is generated. Selecting TSC deadline mode can be done by using mode 0b10 instead of 0b00/0b01 in the timer LVT register.
```

Comment:
use inline code for the binary number? or is ok this way?? (just a question)

Reply:
In terms of presentation it's probably nicer to use inline code for constant values like this, I think you're right :)